### PR TITLE
Fix current test failure

### PR DESCRIPTION
**Pull request by nimec01**
The current implementation of `validBoundsSynTree` can still generate invalid `SynTreeConfig`s. The test cases did not reveal that at my end. I fixed this now.

### CODE REVIEW

**Comment by nimec01** [comment]
There also seems to be problems with `generateLegalPropositionInst`.

**Comment by nimec01** [comment]
Might be a good option to revert #111 as I'm revealing more and more problems.

**Comment by jvoigtlaender** [comment]
I wouldn't be so pessimistic. Let's look at one problem after the other (though from my side that will mean next week).

I guess a lot of the things were simply not really exhaustively tested before, since the "config generators" were a bit naive/narrow and thus not really exploring the full space of what could be input to the task generators. I find it valuable to find these problems now.

**Comment by jvoigtlaender** [comment]
The new https://github.com/fmidue/logic-tasks/blob/1cfc4bdd43b7eeccaae4403ae8af6d0b544f82e5/test/SynTreeSpec.hs#L44 looks good to me.

The "new" problems are maybe not really about `generateLegalPropositionInst` per se. At least, the latest commit on master (f13357fb43759650bacc9d81bc7553527d9724a9) seems to have led to a timeout somewhere in: https://github.com/fmidue/logic-tasks/blob/f13357fb43759650bacc9d81bc7553527d9724a9/test/LegalPropositionSpec.hs#L43-L72

So, it might well be that `validBoundsSynTree` itself is looping in some freak corner cases. My suggestion is to first test this generator more extensively. Possibly by amending https://github.com/fmidue/logic-tasks/blob/f13357fb43759650bacc9d81bc7553527d9724a9/test/SynTreeSpec.hs#L62-L64 locally by using `modifyMaxSuccess` like this:
```haskell
    it "validBoundsSynTree should generate a valid config" $
      modifyMaxSuccess (const 1000) $ 
      forAll validBoundsSynTree $ \synTreeConfig ->
        isJust $ runIdentity $ evalLangM (checkSynTreeConfig synTreeConfig :: LangM Maybe)
```
or similarly.

**Comment by nimec01** [comment]
Using `modifyMaxSuccess (const 10000)` did not change anything for me. Tests are still passing for me.

**Comment by nimec01** [comment]
I was able to grab some SynTreeConfigs that led into non-termination:
```haskell
SynTreeConfig {
minNodes = 4, 
maxNodes = 6, 
minDepth = 3, 
maxDepth = 4, 
usedLiterals = "BDHIKMOQSWXZ", 
atLeastOccurring = 2, 
allowArrowOperators = True, 
maxConsecutiveNegations = 1, 
minUniqueBinOperators = 2
}


SynTreeConfig {
minNodes = 4, 
maxNodes = 10, 
minDepth = 3, 
maxDepth = 4, 
usedLiterals = "AEFJKMORUVWYZ", 
atLeastOccurring = 1, 
allowArrowOperators = True, 
maxConsecutiveNegations = 1, 
minUniqueBinOperators = 2
}

SynTreeConfig {
minNodes = 4, 
maxNodes = 4, 
minDepth = 3, 
maxDepth = 3, 
usedLiterals = "ABCDEFGHIJKLMOQRTUVWYZ", 
atLeastOccurring = 2, 
allowArrowOperators = True, 
maxConsecutiveNegations = 2, 
minUniqueBinOperators = 2
}
```

**Comment by nimec01** [comment]
> For the three `SynTreeConfig`s you give above, does `genSynTree` fail to terminate (sometimes)?

I think so. At least `display synTree` does not terminate.

**Comment by jvoigtlaender** [comment]
Can you check whether the following code produces output?
```haskell
  let nodes = 4 -- then test with 5, and with 6
  in syntaxShape nodes 4 True True
            `suchThat` \synTree ->
              checkAtLeastOccurring synTree &&
              checkMinUniqueOps synTree &&
              consecutiveNegations synTree <= 1)
  where checkAtLeastOccurring synTree = fromIntegral (length (collectLeaves synTree)) >= 2
        checkMinUniqueOps synTree = numOfUniqueBinOpsInSynTree synTree >= 2
```

**Comment by jvoigtlaender** [comment]
Oh, I guess I see that it won't produce output for `nodes = 4`, because there is no tree with just `4` nodes but at least `2` binary operators.

So, another check is needed to prevent such a configuration in the first place.

**Comment by jvoigtlaender** [comment]
A partial fix (helping with the first of the configs you reported) should be to move 
https://github.com/fmidue/logic-tasks/blob/f13357fb43759650bacc9d81bc7553527d9724a9/src/Trees/Generate.hs#L54C15-L54C43
to the "next outer `suchThat` level".

That should definitely be done.

But it won't help with the case
```haskell
SynTreeConfig {
minNodes = 4, 
maxNodes = 4, 
minDepth = 3, 
maxDepth = 3, 
usedLiterals = "ABCDEFGHIJKLMOQRTUVWYZ", 
atLeastOccurring = 2, 
allowArrowOperators = True, 
maxConsecutiveNegations = 2, 
minUniqueBinOperators = 2
}
```
because there the `nodes` will always end up being `4`, even with "retrying".

**Comment by jvoigtlaender** [comment]
To address the latter, I think an additional config check is needed that makes sure `maxNodes` is large enough to really admit `minUniqueBinOperators` binary nodes.

**Comment by jvoigtlaender** [comment]
The only case yet unaccounted for by my above explanation would be this one from above:
```haskell
SynTreeConfig {
  minNodes = 6, 
  maxNodes = 6, 
  minDepth = 3, 
  maxDepth = 3, 
  usedLiterals = "ABCHIJKLMNRUY", 
  atLeastOccurring = 1, 
  allowArrowOperators = False, 
  maxConsecutiveNegations = 1, 
  minUniqueBinOperators = 2}
```

**Comment by jvoigtlaender** [comment]
In that case, the problem could be the `formulas = 7` setting in the "surrounding" `LegalPropositionConfig` being too big to satisfy this: https://github.com/fmidue/logic-tasks/blob/f13357fb43759650bacc9d81bc7553527d9724a9/src/Tasks/LegalProposition/Quiz.hs#L25-L38

That is, there possibly aren't 7 "different enough" trees with exactly `6` nodes and depth `3`.

Again, identifying this cause should allow to strengthen the config checker to weed out such cases.

**Comment by jvoigtlaender** [comment]
About this:
> To address the latter, I think an additional config check is needed that makes sure `maxNodes` is large enough to really admit `minUniqueBinOperators` binary nodes.

I now think the relevant check should actually prevent `minNodes <= 2 * minUniqueBinOperators`.

And the "moving out" of `checkMinUniqueOps synTree` to the outer `suchThat` level is then probably not even needed.

**Comment by nimec01** [comment]
> That is, there possibly aren't 7 "different enough" trees with exactly `6` nodes and depth `3`.
> 
> Again, identifying this cause should allow to strengthen the config checker to weed out such cases.

How does it determine whether two `SynTree`s are similar? Or more specifically: How can I interpret the following code?

```haskell
similarTree :: Eq o => SynTree o c -> SynTree o c -> Bool
similarTree t1 t2 = void t1 == void t2
```

What would be a fitting guard for the config?

File: test/LegalPropositionSpec.hs [hunk 1]
```diff
@@ -59,29 +59,29 @@ spec = do
...
         it "at least creates actual formula symbols" $
-            within timeout $ forAll validBoundsSynTree $ \SynTreeConfig {..} ->
+            within timeout $ forAll validBoundsSynTree $ \synTreeConfig@SynTreeConfig {..} ->
                 forAll
                   (genSynTree synTreeConfig) $ \synTree ->
                       forAll (deleteSpaces <$> illegalDisplay synTree) $
                       all (\c -> c `elem` "()∧∨¬<=>" || isLetter c)
         it "the string after illegalDisplay cannot be parsed" $
-            within timeout $ forAll validBoundsSynTree $ \SynTreeConfig {..} ->
+            within timeout $ forAll validBoundsSynTree $ \synTreeConfig@SynTreeConfig {..} ->
```
Comment (jvoigtlaender) [inline test/LegalPropositionSpec.hs:68]: Is the `@SynTreeConfig {..}` part actually still needed? It seems that no access to the fields actually happens below. (Might also be relevant at other places.)

File: test/SynTreeSpec.hs [hunk 2]
```diff
@@ -35,15 +35,15 @@ validBoundsSynTree = do
...
+  minNodes <- choose (max 3 (minAmountOfUniqueAtoms * 2), 60) `suchThat` \minNodes' -> maxConsecutiveNegations /= 0 || odd minNodes'
   let minDepth = 1 + floor (logBase (2 :: Double) $ fromIntegral minNodes)
   let minMaxDepth = max (maxConsecutiveNegations + 1) minDepth
   maxDepth <- choose (minMaxDepth,max (minMaxDepth+ 3) 10)
   maxNodes <- choose (minNodes, maxNodesForDepth maxDepth) `suchThat`
     \maxNodes' -> (maxConsecutiveNegations /= 0 || odd maxNodes')
       && maxDepth <= maxDepthForNodes maxConsecutiveNegations maxNodes'
   let availableBinOpsCount = if allowArrowOperators then fromIntegral $ length [minBound .. maxBound :: BinOp] else 2
-  minUniqueBinOperators <- choose (1, min (minDepth - 1) availableBinOpsCount)
+  minUniqueBinOperators <- choose (1, min (minDepth - 1) (min availableBinOpsCount ((minNodes `div` 2) - 1)))
```
Comment (jvoigtlaender) [inline test/SynTreeSpec.hs:46]: Nested `min`s are probably nicer rewritten using `minimum` on a list.

Also, the current generator here now would "outlaw" `minNodes = 7` and `minUniqueBinOperators = 3`, even though that is a setting that in principle should be okay.

**Comment by jvoigtlaender** [comment]
> How can I interpret the following code?

That code takes `t1 :: SynTree o c` and `t2 :: SynTree o c`, turns both into a `SynTree o ()` each, by replacing each leaf label by `()`, then compares the resulting structures.

So what it is really doing is compare whether the two trees have exactly the same "shape" (also taking into account which operators are labelled at the inner nodes).

**Comment by jvoigtlaender** [comment]
With
```haskell
SynTreeConfig {
  minNodes = 6, 
  maxNodes = 6, 
  minDepth = 3, 
  maxDepth = 3, 
  usedLiterals = "ABCHIJKLMNRUY", 
  atLeastOccurring = 1, 
  allowArrowOperators = False, 
  maxConsecutiveNegations = 1, 
  minUniqueBinOperators = 2}
```
there are only four possible shapes in that sense:
- `And (Or () ()) (Not ())`
- `And (Not ()) (Or () ())`
- `Or (And () ()) (Not ())`
- `Or (Not ()) (And () ())`

Hence, `formulas = 7` is too much.

File: test/LegalPropositionSpec.hs (hunk 1 above)
Comment (nimec01) [reply test/LegalPropositionSpec.hs]: No, I removed it.

File: test/SynTreeSpec.hs (hunk 2 above)
Comment (nimec01) [reply test/SynTreeSpec.hs]: I adjusted it.

**Comment by nimec01** [comment]
Most of the mentioned changes should be implemented now. There is a problem with SuperfluousBrackets which I'm investigating now.

File: test/LegalPropositionSpec.hs [hunk 3]
```diff
@@ -20,20 +20,20 @@ import Tasks.LegalProposition.Quiz (generateLegalPropositionInst)
...
 import Control.Monad.Output (LangM)
 import Data.Maybe (isJust)
 import Control.Monad.Identity (Identity(runIdentity))
 import Control.Monad.Output.Generic (evalLangM)
+import Tasks.LegalProposition.Helpers (formulaAmount)
 
 validBoundsLegalProposition :: Gen LegalPropositionConfig
 validBoundsLegalProposition = do
+    formulas <- choose (1, 15)
     syntaxTreeConfig@SynTreeConfig {..}  <- validBoundsSynTree `suchThat` ((15 <=) . minNodes)
```
Comment (jvoigtlaender) [inline test/LegalPropositionSpec.hs:35]: Maybe undo the `15 <=` here now?

File: test/LegalPropositionSpec.hs [hunk 4]
```diff
@@ -20,20 +20,20 @@ import Tasks.LegalProposition.Quiz (generateLegalPropositionInst)
...
-    let leaves = maxLeavesForNodes maxNodes
-    formulas <- choose (1, min 15 $ if allowArrowOperators then 4 else 2 ^ (maxNodes - leaves))
```
Comment (jvoigtlaender) [inline test/LegalPropositionSpec.hs:28]: Oh, the previous problem possibly was partly due to a mis-bracketing here. Quite likely `if allowArrowOperators then 4 else 2 ^ (maxNodes - leaves)` was actually meant to be `(if allowArrowOperators then 4 else 2) ^ (maxNodes - leaves)`.

**Comment by nimec01** [comment]
SuperfluousBracketsSpec had some test cases being stuck sometimes. I would guess that this is happening due to the combination of 
```haskell
`suchThat` sameAssociativeOperatorAdjacent
```
used in `generateSuperfluousBracketsInst` and a low `maxNodes` / `maxDepth` setting. 

There might be a better way, but I changed it to only use configs with at least 8 nodes.

**Comment by jvoigtlaender** [comment]
Such repairs by letting certain counts have a minimal value (in the config generators in the test suite) should then probably also make their way into the config checkers (on lecturer-provided configs).

After all, it doesn't really make sense to systematically test fewer cases than may appear in actual usage. Or, put differently, the validity checkers (being used and tested) should be as strong as what constraints are added to the test case generators for making tests go through. Otherwise, we are deliberately saying "let's not test this case, because the test sometimes fails, but let's still allow the lecturer to make a configuration with this known-to-be-causing-failures characteristics". At least if the additional constraint is not too far-fetched for teaching use cases.
